File: Source/Composition/CloseBBcodeTagCommand.m
//  CloseBBcodeTagCommand.m
//
//  Copyright 2014 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "CloseBBcodeTagCommand.h"

@interface CloseBBcodeTagCommand ()

@property (assign, nonatomic) BOOL enabled;

@end

@implementation CloseBBcodeTagCommand

- (instancetype)initWithTextView:(UITextView *)textView
{
    if ((self = [super init])) {
        _textView = textView;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textDidChange:) name:UITextViewTextDidChangeNotification object:textView];
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)textDidChange:(NSNotification *)note
{
    NSString *textContent = [self.textView.text substringToIndex:self.textView.selectedRange.location];
    self.enabled = !![self getCurrentlyOpenTag:textContent];
}

- (void)execute
{
    NSString *textContent = [self.textView.text substringToIndex:self.textView.selectedRange.location];
    
    if ([self hasOpenCodeTag:textContent]) {
        [self.textView insertText:@"[/code]"];
        return;
    }
    
    NSString *openTag = [self getCurrentlyOpenTag:textContent];
    if (openTag) {
        [self.textView insertText:[NSString stringWithFormat:@"[/%@]", openTag]];
    }
}

/*
 * Insert the appropriate closing tag, if any.
 *
 * First, scan backwards for [code]. If so, and there's no [/code] between there and here, then
 * the insertion is always [/code] (bbcode within [code] isn't interpreted).
 *
 * Scan backwards looking for [.
 * - If we find a [/tag], scan backwards for its [tag] and continue search from there.
 * - If we find [tag], trim =part if any, and insert [/tag].
 *
 * XXX should have a list of bbcode tags, and only consider those?
 */

/* Tests:
 * "[code] [b]"                -> TRUE
 * "[code] [b] [/code]"        -> FALSE
 * "[code] [b] [/code][code]"  -> TRUE
 * "[code=cpp] [b]"            -> TRUE
 * "[/code]"                   -> FALSE
 * "[codemonkey] [b]"          -> FALSE
 * "[code][codemonkey]"        -> TRUE
 */
- (BOOL)hasOpenCodeTag:(NSString *)content
{
    NSRange codeRange = [content rangeOfString:@"[code" options:NSBackwardsSearch];
    if (codeRange.location == NSNotFound || NSMaxRange(codeRange) >= content.length) {
        return NO;
    }
    
    // If it's a false alarm like [codemonkey], keep looking.
    unichar nextChar = [content characterAtIndex:NSMaxRange(codeRange) /* [code */];
    if (![TagNameTerminators() characterIsMember:nextChar]) {
        return [self hasOpenCodeTag:[content substringToIndex:codeRange.location]];
    }
    
    // Is this still open?
    return [[content substringFromIndex:codeRange.location] rangeOfString:@"[/code]"].location == NSNotFound;
}

/*
 * Tests:
 * "[b][i]"              -> "i"
 * "[b][i][/i]"          -> "b"
 * "[b][/b]"             -> nil
 * "[url=foo]"           -> "url"
 * "[url=foo][b][i][/b]" -> "url"
 * "["                   -> "nil"
 * "[foo][/x"            -> "foo"
 * "[foo attr]"          -> "foo"
 * "[code][b]"           -> "code"
 * "[b][code][/code]"    -> "b"
 * "[list][*]"           -> "list"
 */
- (NSString *)getCurrentlyOpenTag:(NSString *)content
{
    // Find start of preceding tag (opener or closer).
    NSUInteger startingBracket = [content rangeOfString:@"[" options:NSBackwardsSearch].location;
    if (startingBracket == NSNotFound) {
        return nil;
    }
    
    if (startingBracket >= content.length - 1) {
        // Incomplete tag, keep going.
        return [self getCurrentlyOpenTag:[content substringToIndex:startingBracket]];
    }
    
    // If it's a closer, find its opener.
    if ([content characterAtIndex:(startingBracket + 1)] == '/') {
        NSRange tagRange = [[content substringFromIndex:startingBracket] rangeOfString:@"]"];
        if (tagRange.location == NSNotFound) {
            // Not a proper tag, keep searching backwards.
            return [self getCurrentlyOpenTag:[content substringToIndex:startingBracket]];
        }
        
        tagRange = NSMakeRange(startingBracket + 2, tagRange.location - 2);
        NSString *tagname = [content substringWithRange:tagRange];
        
        NSUInteger openerLocation =
        [content rangeOfString:
         [NSString stringWithFormat:@"[%@]", tagname] options:NSBackwardsSearch].location;
        
        if (openerLocation == NSNotFound) {
            // Might be [tag=attr]
            openerLocation =
            [content rangeOfString:
             [NSString stringWithFormat:@"[%@=", tagname] options:NSBackwardsSearch].location;
        }
        
        if (openerLocation == NSNotFound) {
            // Might be [tag attr=val]
            openerLocation =
            [content rangeOfString:
             [NSString stringWithFormat:@"[%@ ", tagname] options:NSBackwardsSearch].location;
        }
        
        if (openerLocation == NSNotFound) {
            // Never opened, keep searching backwards from the starting bracket.
            return [self getCurrentlyOpenTag:[content substringToIndex:startingBracket]];
        }
        
        // Now that we've matched [tag]...[/tag], keep looking back for an outer [tag2] that
        // might still be open.
        return [self getCurrentlyOpenTag:[content substringToIndex:openerLocation]];
    }
    
    // We have an opener! Find the end of the tag name.
    NSRange tagRange = [content rangeOfCharacterFromSet:TagNameTerminators()
                                                options:0
                                                  range:NSMakeRange(startingBracket + 1, content.length - startingBracket - 1)];
    if (tagRange.location == NSNotFound) {
        // Malformed, keep looking.
        return [self getCurrentlyOpenTag:[content substringToIndex:startingBracket]];
    }
    
    tagRange.length--; // Omit the ] or =;
    NSString *tagName = [content substringWithRange:NSMakeRange(startingBracket + 1, tagRange.location - startingBracket - 1)];
    if ([tagName isEqualToString:@"*"]) {
        return [self getCurrentlyOpenTag:[content substringToIndex:startingBracket]];
    }
    
    return tagName;
}

static NSCharacterSet * TagNameTerminators(void)
{
    return [NSCharacterSet characterSetWithCharactersInString:@"]= "];
}

@end

File: App/Composition/CloseBBcodeTagCommand.swift
//  CloseBBcodeTagCommand.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import UIKit

/// Auto-closes the nearest open BBcode tag in a text view.
final class CloseBBcodeTagCommand: NSObject {

    /// Whether the command can execute. KVO-compliant.
    @objc dynamic private(set) var enabled = false
    
    private let textView: UITextView
    
    init(textView: UITextView) {
        self.textView = textView
        super.init()
        
        updateEnabled()
        
        NotificationCenter.default.addObserver(self, selector: #selector(textDidChange), name: .UITextViewTextDidChange, object: textView)
    }
    
    @objc private func textDidChange(notification: Notification) {
        updateEnabled()
    }
    
    private var textContent: Substring {
        return textView.text[..<String.Index(encodedOffset: textView.selectedRange.location)]
    }
    
    private func updateEnabled() {
        enabled = getCurrentlyOpenTag(textContent) != nil
    }
    
    /// Closes the nearest open BBcode tag.
    func execute() {
        if hasOpenCodeTag(textContent) {
            textView.insertText("[/code]")
            return
        }
        
        if let openTag = getCurrentlyOpenTag(textContent) {
            textView.insertText("[/\(openTag)]")
        }
    }
}

/*
 * Insert the appropriate closing tag, if any.
 *
 * First, scan backwards for [code]. If so, and there's no [/code] between there and here, then
 * the insertion is always [/code] (bbcode within [code] isn't interpreted).
 *
 * Scan backwards looking for [.
 * - If we find a [/tag], scan backwards for its [tag] and continue search from there.
 * - If we find [tag], trim =part if any, and insert [/tag].
 *
 * XXX should have a list of bbcode tags, and only consider those?
 */

/* Tests:
 * "[code] [b]"                -> true
 * "[code] [b] [/code]"        -> false
 * "[code] [b] [/code][code]"  -> true
 * "[code=cpp] [b]"            -> true
 * "[/code]"                   -> false
 * "[codemonkey] [b]"          -> false
 * "[code][codemonkey]"        -> true
 */
internal func hasOpenCodeTag(_ text: Substring) -> Bool {
    guard let codeRange = text.range(of: "[code", options: .backwards), codeRange.upperBound < text.endIndex else {
        return false
    }
    
    // If it's a false alarm like [codemonkey], keep looking.

    let nextCharacter = text[codeRange.upperBound]
    if nextCharacter.unicodeScalars.count == 1, !tagNameTerminators.contains(nextCharacter.unicodeScalars.first!) {
        return hasOpenCodeTag(text[..<codeRange.lowerBound])
    }
    
    // Is this still open?
    return text[codeRange.lowerBound...].range(of: "[/code]") == nil
}

/*
 * Tests:
 * "[b][i]"              -> "i"
 * "[b][i][/i]"          -> "b"
 * "[b][/b]"             -> nil
 * "[url=foo]"           -> "url"
 * "[url=foo][b][i][/b]" -> "url"
 * "["                   -> nil
 * "[foo][/x"            -> "foo"
 * "[foo attr]"          -> "foo"
 * "[code][b]"           -> "code"
 * "[b][code][/code]"    -> "b"
 * "[list][*]"           -> "list"
 */
internal func getCurrentlyOpenTag(_ text: Substring) -> Substring? {
    // Find start of preceding tag (opener or closer).
    guard let startingBracket = text.range(of: "[", options: .backwards) else { return nil }
    
    if startingBracket.upperBound == text.endIndex {
        // Incomplete tag, keep going.
        return getCurrentlyOpenTag(text[..<startingBracket.lowerBound])
    }
    
    // If it's a closer, find its opener.
    if text[startingBracket.upperBound] == "/" {
        guard let tagEnd = text[startingBracket.lowerBound...].range(of: "]") else {
            // Not a proper tag, keep searching backwards.
            return getCurrentlyOpenTag(text[..<startingBracket.lowerBound])
        }

        let afterSlash = text.index(after: startingBracket.upperBound)
        let tagName = text[afterSlash ..< tagEnd.lowerBound]
        guard let opener = text.range(of: "[\(tagName)]", options: .backwards)
            // Might be [tag=attr]
            ?? text.range(of: "[\(tagName)=", options: .backwards)
            // Might be [tag attr=val]
            ?? text.range(of: "[\(tagName) ", options: .backwards) else
        {
            // Never opened, keep searching backwards from the starting bracket.
            return getCurrentlyOpenTag(text[..<startingBracket.lowerBound])
        }

        // Now that we've matched [tag]...[/tag], keep looking back for an outer [tag2] that might still be open.
        return getCurrentlyOpenTag(text[..<opener.lowerBound])
    }
    
    // We have an opener!
    guard text[startingBracket.lowerBound...].range(of: "]") != nil else {
        // User is still typing the tag, we're done here.
        return nil
    }
    
    // Find the end of the tag name.
    guard let terminator = text.rangeOfCharacter(from: tagNameTerminators, options: [], range: startingBracket.upperBound ..< text.endIndex) else {
        // Malformed, keep looking.
        return getCurrentlyOpenTag(text[..<startingBracket.lowerBound])
    }

    let tagName = text[startingBracket.upperBound ..< terminator.lowerBound]
    if tagName == "*" {
        return getCurrentlyOpenTag(text[..<startingBracket.lowerBound])
    }
    
    return tagName
}

private let tagNameTerminators = CharacterSet(charactersIn: "]= ")
